Add wildcard_match_arm lint

Aehmlo opened the pull request:
This lint prevents using a wildcard in a match arm. Implemented as a restriction currently, because this is pretty much an edge case. See #3649 for more information.

Didn't add any tests because I wasn't sure how, but if someone wants to point me in the right direction, I'd be happy to!

flip1995 commented on clippy_lints/src/matches.rs:
```diff
@@ -438,6 +459,22 @@ fn check_wild_err_arm(cx: &LateContext<'_, '_>, ex: &Expr, arms: &[Arm]) {
+                span_note_and_lint(cx,
+                    MATCH_WILD,
+                    arm.pats[0].span,
+                    "Wildcard match will miss any future added variants.",
```
NIT: 
```suggestion
                    "wildcard match will miss any future added variants",
```

flip1995 commented on clippy_lints/src/matches.rs:
````diff
@@ -187,6 +187,25 @@ declare_clippy_lint! {
+/// }
+/// ```
+declare_clippy_lint! {
+    pub MATCH_WILD,
````
```suggestion
    pub WILDCARD_MATCH_ARM,
```
lint names should build a sentence (more or less) if they're put after an "allow": #2845 
"allow wildcard match arm" is more a sentence than "allow match wild".

Aehmlo commented:
Update: I've made some changes (including the above), but it'll be a few days until I can finish implementing the tests. I'll push everything once that's done.

Aehmlo commented:
Wrong button! I think this should be good to go now.

flip1995 commented:
Thanks! 

Only a `rustfmt` run is missing, then this will be good to go.
https://travis-ci.com/rust-lang/rust-clippy/jobs/173071805#L1387-L1415

Aehmlo commented:
Ran `rustfmt` and pushed, but it seems that there are some other nightly-related breakages that popped up.

flip1995 commented:
Thanks! That should be fixed by now.

@bors r+

bors commented:
:pushpin: Commit fefe55eb7ea16624e686fcb488611c8b24d087c8 has been approved by `flip1995`

<!-- @bors r=flip1995 fefe55eb7ea16624e686fcb488611c8b24d087c8 -->

bors commented:
:hourglass: Testing commit fefe55eb7ea16624e686fcb488611c8b24d087c8 with merge a70d648240e4428e36a25e6a81be1716a98a56d0...

bors commented:
:broken_heart: Test failed - [checks-travis](https://travis-ci.com/rust-lang/rust-clippy/builds/98947545)

flip1995 commented:
It seems that a formatting of the tests is missing.

https://travis-ci.com/rust-lang/rust-clippy/jobs/173709864#L1383-L1447

A `cargo fmt --all` inside the crate root should format everything

Aehmlo commented:
Neither `cargo fmt --all` nor `cargo +nightly fmt --all` was working to format the tests, so I just ran `rustfmt` on the file directory. Regardless, it looks like the build is now passing!

orium commented:
I think it would be better just to emit a lint in cases of an enum. So

```rust
match 3 {
    4 => println!(),
    _ => println!(),
};
```

would not emit a warning.

Note also that it would need to check for matching in an enum on any nesting level.  I.e.

```rust
match Some(Color::Red) {
    Some(Color::Blue) => println!(),
    Some(_) => println!(),
    None => println!(),
};
```

would emit a lint.

phansch commented:
@bors r- pending decision on linting non-enum matches

Aehmlo commented:
I agree re: linting only on enums. I've changed the behavior and updated the lint name accordingly. I'm not familiar enough with compiler internals to implement the nested check yet, so I added a note of this limitation to the known issues. I may try to resolve this in the future if I have time, or someone else can, but I won't be able to do it in the near future.

Thanks for the insightful feedback, @orium!

bors commented:
:pushpin: Commit 97c7d2818aad2a9a400ca954e3db61a4c3d5d781 has been approved by `phansch`

<!-- @bors r=phansch 97c7d2818aad2a9a400ca954e3db61a4c3d5d781 -->

bors commented:
:hourglass: Testing commit 97c7d2818aad2a9a400ca954e3db61a4c3d5d781 with merge c254711eb36ad05d24aa1e7a2b47b9dfb4b59016...

Aehmlo commented:
Updated lint count; should actually work next time.

matthiaskrgr commented:
@bors retry

matthiaskrgr commented:
@bors ping

bors commented:
:sleepy: I'm awake I'm awake

matthiaskrgr commented:
@bors r=phansch

bors commented:
:pushpin: Commit 7fa50fb3fe98f3c6f837e95e6d13810c68ceaf74 has been approved by `phansch`

<!-- @bors r=phansch 7fa50fb3fe98f3c6f837e95e6d13810c68ceaf74 -->

bors commented:
:hourglass: Testing commit 7fa50fb3fe98f3c6f837e95e6d13810c68ceaf74 with merge 6ce78d1257ac6fd77f245730fcfbadd536a173eb...

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust-clippy/builds/99042815), [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/rust-clippy/builds/21983975)
Approved by: phansch
Pushing 6ce78d1257ac6fd77f245730fcfbadd536a173eb to master...
